eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java

```diff
@@ -165,185 +161,230 @@ private void getOrCreateClusteredVariantAccessions(List<? extends SubmittedVaria
-        List<GetOrCreateAccessionWrapper<IClusteredVariant, String, Long>> accessionWrappers =
-                clusteredService.getOrCreate(clusteredVariantNewRS);
-        long newAccessions = accessionWrappers.stream().filter(GetOrCreateAccessionWrapper::isNewAccession).count();
-        clusteringCounts.addClusteredVariantsCreated(newAccessions);
+    private void insertAllEntriesInDB(List<ClusteredVariantEntity> clusteredVariantEntities,
+                                      Map<String, SubmittedVariantOperationEntity> mergeSVOE,
+                                      Map<Long, SubmittedVariantOperationEntity> rsSplitSVOE) {
 
-        return accessionWrappers;
-    }
+        mongoTemplate.insert(clusteredVariantEntities, ClusteredVariantEntity.class);
 
-    private List<ClusteredVariantOperationEntity> getClusteredVariantOperations(List<GetOrCreateAccessionWrapper<IClusteredVariant, String, Long>> accessionWrappers,
-                                                                                Map<String,SubmittedVariantEntity> submittedVariantEntitiesGeneratingNewRS){
-        List<ClusteredVariantOperationEntity> clusteredVariantOperationEntities = new ArrayList<>();
-        for(GetOrCreateAccessionWrapper<IClusteredVariant,String,Long> accessionWrapper : accessionWrappers) {
-            Long originalClusteredAccession = submittedVariantEntitiesGeneratingNewRS.get(accessionWrapper.getHash()).getClusteredVariantAccession();
-            ClusteredVariantOperationEntity clusteredVariantOperationEntity = new ClusteredVariantOperationEntity();
-            clusteredVariantOperationEntity.fill(EventType.RS_SPLIT,
-                                                 originalClusteredAccession,
-                                                 accessionWrapper.getAccession(),
-                                                 new StringBuilder().append("Due to RS split new accession id ")
-                                                                    .append(accessionWrapper.getAccession())
-                                                                    .append(" created for remapped accession ")
-                                                                    .append(originalClusteredAccession)
-                                                                    .toString(),
-                                                 null);
-            clusteredVariantOperationEntities.add(clusteredVariantOperationEntity);
+        for (Map.Entry<String, SubmittedVariantOperationEntity> entry : mergeSVOE.entrySet()) {
+            Query querySubmitted = query(where("eventType").is(EventType.RS_MERGE_CANDIDATES)
+                    .and("accession").is(entry.getValue().getAccession()));
+            Update update = new Update();
+            update.set("inactiveObjects", entry.getValue().getInactiveObjects());
+            mongoTemplate.upsert(querySubmitted, update, SubmittedVariantOperationEntity.class);
```

Comment:
I am wondering what happens when we don't already have an entry for a given accession? We seem to be only inserting the inactiveObjects array but not the rest of the document in that case.

Reply:
Also, unfortunately, upserts to a sharded collection must contain the shard key (_id in this case) in the query field :(

Reply:
> I am wondering what happens when we don't already have an entry for a given accession? We seem to be only inserting the inactiveObjects array but not the rest of the document in that case.

As rest of the fields are part of the query, they are also getting inserted along with inactiveObjects. I never realized, this could be problem as it was working fine and I could see all the entries in the DB. 


> Also, unfortunately, upserts to a sharded collection must contain the shard key (_id in this case) in the query field :(.

When we are inserting the entry first time, we don't have any id with us. Do we need to separate the entries in 2 different parts (first time inserts, other as upserts with the id) ? or is there any other better way to handle this ? (assuming we are not able to generate the correct id and put it ourselves for the first time entry)

Reply:
> rest of the fields are part of the query
I don't think "reason" attribute is part of the query. 
> When we are inserting the entry first time, we don't have any id with us.
In that case, we should just do separate inserts and updates separately.

Reply:
You can create a sharded collection in PROD in a test database and check the shard key requirement for upserts.

Reply:
separated inserts and upserts statements